### ISSUE TITLE
[Proposal] Array Items on Separate Lines Fixer

### ISSUE THREAD

**Issue by vinkla** [PHP-CS-Fixer/PHP-CS-Fixer#1217]
I often find myself putting array items on separate lines. Specially when there are a lot of items in the arrays. Maybe we could add a fixer for that? Maybe it exists already? Example below:

``` php
// Before
protected $dates = ['deleted_at', 'ended_at', 'started_at'];

// After
protected $dates = [ 
    'deleted_at', 
    'ended_at', 
    'started_at'
];
```

What do you think?

**Comment by GrahamCampbell** [comment]
Do people always want this everywhere in there code though? This fixer would be very annoying for the following:

```php
$factory->render('foo', ['bar'], false);
```

as it would end up:

```php
$factory->render('foo', [
    'bar'
], false);
```

**Comment by vinkla** [comment]
@GrahamCampbell Thats true. Maybe only for variables?

**Comment by tlenex** [comment]
+1 for variables

**Comment by Anahkiasen** [comment]
I'd love to see this in but as mentioned, only for properties and variables. Rest I prefer to do by hand

**Comment by TomasVotruba** [comment]
I'm looking for similar fixer, probably configurable.

```php
// no keys, one line
$elements = [$path, $name]; 

// keys
$elements = [ 
    'file' => $path,
    'label' => $name
];
```

I'd appreciate starting points on fixer architecture with arrays, so I could make this happen.

**Comment by TomasVotruba** [comment]
If anyone is interested in this fixer, I made similar prototype: https://github.com/Symplify/CodingStandard/blob/master/README.md#indexed-php-arrays-should-have-1-item-per-line

**Comment by keradus** [comment]
feel free to send a PR

**Comment by TomasVotruba** [comment]
Not in my prirority list Now.

**Comment by Soullivaneuh** [comment]
> Do people always want this everywhere in there code though?

Well, don't use it. :trollface: 

More seriously, we should add a `min_items` option, like for this eslint rule: https://eslint.org/docs/rules/array-element-newline

BTW, we may done the same fixer for function signature parameters.

**Comment by ikari7789** [comment]
> Do people always want this everywhere in there code though?

I think just simply adding it as an option adds no harm. Can always improve it with configurable settings after the fact. Maybe you don't want it everywhere in the code, but maybe some people do. This proposal seems to have been sitting here now for three years without any movement, but I'd love to see something come of this.

**Comment by taylankasap** [comment]
Is adding a configuration to `WhitespaceAfterCommaInArrayFixer` instead of creating a new fixer good idea? Even the name fits, since newline is also a whitespace.

**Comment by goyote** [comment]
has anybody found a solution for this?

**Comment by madebycaliper** [comment]
Would also love a configurable version of this rule. Maybe based on max line length? (I'm not a CS Fixer expert, sorry if its a dumb idea)
``` php
// Remains Untouched
$foo = ['bar' => $bar, 'biz' => $biz];

// Long Assoc Arrays are wrapped and stacked
$foo = ['bar' => $bar, 'biz' => $biz, 'anotherSweetKey' => $totallRadVar, 'longEnoughToBreakLineLengthMax' => $something ];

$foo = [
    'bar' => $bar,
    'biz' => $biz,
    'anotherSweetKey' => $totallRadVar,
    'longEnoughToBreakLineLengthMax' => $something
];
```
 
If anyone knows whether this exists already, I'd love a clue on which rule to use. Thanks.

**Comment by julienfalque** [comment]
IMO this should be a combination of two distinct rules:
- a general rule that chops down long lines when possible (parameters list, array items, ...)
- one that forces arrays to have all items on the same line or one item per line (like we already do for function arguments).

Both rules don't exist at the moment.

**Comment by charlesdeb** [comment]
This would be handy for WP developers since for more than 1 item in an array, the WP PHP coding standard calls for new elements on a  new line.

See https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/#indentation

**Comment by Haythamasalama** [comment]
+1

**Comment by kubawerlos** [comment]
Hi folks, you can do better than bumping the topic - helping to have this feature by raising PR with its implementation.

**Comment by alexander-zierhut** [comment]
> 1. a general rule that chops down long lines when possible (parameters list, array items, ...)
> 1. one that forces arrays to have all items on the same line or one item per line (like we already do for function arguments).
After some digging around for examples I came up with two more possible rules, but I am not sure how to properly define them yet.
3. Sometimes you want to use a stacked array, but want to have multiple items in the same row for some rows as you might want to group them semantically. This could be a configuration option to i.e. use stacked arrays, but don't push items into the next row if two or more are already in a given row.
```PHP
$array = [
	$variable1,
	$r, $g, $b,
	$variable2,
];
```
4. Always force values into a stacked array when inside of another array or when are used in the context of a function call as a single parameter.
```PHP
$array = [
	"value1" => $variable1,
	"value2" => [
		"inner1" => 1,
		"inner2" => 2,
	],
	"value3" => "string3",
];

$obj->doSomething([
	"option1" => "value1",
	"option2" => "value2",
]);
```

**Comment by Wirone** [comment]
@alexander-zierhut regarding scenario no. 3: how tool would determine if elements in the same row are there on purpose or because someone did not hit enter? I think it would be impossible to handle it nicely and would only bring maintenance burden.

**Comment by alexander-zierhut** [comment]
@Wirone I agree, maybe someone else can come up with a better idea, but I think it's still worth being an option, which is off by default.

**Comment by adjenks** [comment]
I used this:
https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6315

**Comment by madebycaliper** [comment]
Adding a comment here to prevent this from closing. I think an implementation of this rule would still be very helpful, despite this issue being almost 10 years old without a PR 🤣

**Comment by Wirone** [comment]
@madebycaliper there _IS_ a PR 😉. And the issue wouldn't be closed as I already removed stale label before 😅.

**Comment by soullivaneuh** [comment]
> Do people always want this everywhere in there code though? This fixer would be very annoying for the following:
> 
> ```
> $factory->render('foo', ['bar'], false);
> ```
> 
> as it would end up:
> 
> ```
> $factory->render('foo', [
>     'bar'
> ], false);
> ```

Since latest PHP versions, we can also multi-line function parameters with a coma (`,`) at each line, reducing diff noise and possible conflict.

We MIGHT consider php-cs-fixer managing both, producing something like (for this case):

```php
$factory->render(
    'foo',
    [
        'bar',
    ],
    false,
);
```

What do you think? :thinking:

**Comment by jlecordier** [comment]
Could we take into account the fact that some entries are already on multiple lines, regardless of the length of the array ?

See [my original comment](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6930#issuecomment-2129542709)

Before : 

```php
$tab = [
    'a',
    'b', 'c'
];
```

After :

```php
$tab = [
    'a',
    'b',
    'c'
];
```

**Comment by jg-development** [comment]
plz no stale ;-)

**Comment by adjenks** [comment]
**_MY HANDS ARE TYPING WORDS_**

**Comment by kubawerlos** [comment]
> **_MY HANDS ARE TYPING WORDS_**

**_WILL THESE WORDS TURN INTO A PULL REQUEST WITH THE FIXER?_**

**Comment by adjenks** [comment]
@kubawerlos I was referencing this comic; https://xkcd.com/1296/
I mentioned previously that I myself used this pull request: https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/6315
I feel like the maintainers of PHP-CS-Fixer could simply make use of that request or parts of it if they wanted to close this.

**Comment by jasverix** [comment]
One suggestion here. If an array is single-line, it can stay that way. But if it's partially multi-line, it would be nice if the linter can make it proper multi line.

```
$arr = [
    'one' => 1, 'two' => 2,
    'three' => 3,
    'four' => 4,
];
```
to
```
$arr = [
    'one' => 1,
    'two' => 2,
    'three' => 3,
    'four' => 4,
];
```

**Comment by sebj54** [comment]
beep boop

**Comment by madebycaliper** [comment]
Would still like to see this 👍

**Comment by django23** [comment]
Let's keep this alive.

**Comment by DKhalil** [comment]
Ah ah ah ah stayin alive stayin alive

**Comment by DKhalil** [comment]
There is a PR open already with #9703 so keep this open
